lib/pbench/server/api/resources/users_api.py

```diff
@@ -473,6 +474,10 @@ def delete(self, username):
             # Do not delete if the user is admin
             if not user.is_admin():
                 User.delete(username)
+        except NoResultFound:
+            self.logger.exception(
+                "Delete operation on non-existent user '{}'", username,
+            )
```

Comment:
So we're logging an exception but returning success?

I think that, first, in this case the `__cause__` is pretty obvious, so the `self.logger.exception` is unnecessarily noisy, plus `exception` is implicitly `error` level, and that's inappropriate if we're treating this as success. It may be worth noting (although I'm not convinced since it's probably a user typo or an enormously improbable collision), but it's probably an `info` level at most. Also, it's clearly a user error, and we've more or less adopted the convention of logging those at `warning` level... which I think is probably heavy in this context.

With that said ... at this point it may not be worth another round, unless you need to rebase or Pete wants it squashed pre-merge, so I'm approving. I guess we'll see if others agree. ;-)

Reply:
Yes I think it should an `info`

Reply:
Ah; I hoped you might fix this when you rebased, but that's OK.

Reply:
Ah I need a another rebase

Reply:
Could either of `User.query()` or `user.is_admin()` raise `NoResultFound`?  That is, what happens if `username` doesn't exist in the database?  I assume that control never gets anywhere near the call to `User.delete()`....

Reply:
User.query just return `None` in case no user is found and `is_admin` is an object method it returns True/False we can't call that method if the user is not found. we check if the user is present first and then call `is_admin` does that make sense?

Reply:
If the `username` is not in the database, then (I think...) `verified` will be false, in which case we can abort without worrying about whether the requestor is an admin.  (That is, even if the requestor _is_ an admin, that doesn't change the fact that the target user doesn't exist....)

I feel like the structure of this code is obfuscating its function.  `verify_user()` is really "determine whether the requesting user matches the specified username and return the requesting `user` value", which really isn't very much code.  (That is, it seems like, as things are, the situation would be clearer if the subroutine body were just in-line'd.)  And, then, we have to layer the admin checks on top of that.

I think the code would be better if the subroutine returned a boolean indicating whether the current user had the right to update/destroy the specified user and also, for convenience, the specified user's `User`.  That would encapsulate most of the `is_admin()` checks in one place, and the calling code could do a simple check of the "verified" value.  That is, if the specified user doesn't exist, the caller gets a `False`; if the requestor is an admin, the caller gets a `True`; otherwise, the caller gets a `True` only if the target matches the requestor.  And, we get the target back, which (I think) saves us a database query (i.e., I don't think we need to query the database for the requestor's `User` -- I think that comes from the `Auth` `current_user()` call -- so we only need to query for the target's `User`).  The caller can determine the "user does not exist" case based on a `user == None` return value, and otherwise a `verified == False` means "not authorized".

I don't think it's helpful to reuse the `user` variable -- it initially represents the requestor, but it is later overwritten with the target, and that's potentially confusing.

Also, I suggest reworking the criteria around deleting an admin user:  if a user is not an admin, they can only delete themself; if a user _is_ an admin, they should be able to delete any user _except_ themself.  (If there is only one admin user, "admin", then it won't be able to be deleted.)  (This logic would go in `delete()`, not in `verify_user()`.)

Reply:
Okay I think refactoring `verify_user()` might be an issue for a separate PR, I will create an issue to track. I agree with the delete method update, I almost forgot via rest api we do not allow admin user to get self deleted, we should add that check in the cli as well.

Reply:
Created an issue to track https://github.com/distributed-system-analysis/pbench/issues/2275

Reply:
It just occurred to me that, in the CLI, deleting the admin user is not the catastrophic event that it might be via the API:  the API requires the requestor to be authenticated and authorized, and so deleting the admin user might result in the inability to manage the system (via the API); however, the CLI bypasses all that security, so it would be quite straightforward to re-create the admin user (using the CLI).

So, we probably don't need to have very many safeguards for the CLI.  In fact, we might not _want_ to have too many safeguards:  the CLI won't be generally used -- it's just for set-up and testing -- so being able to do things which the API _cannot_ do might well be a **feature**.  :-)